xmodule/modulestore/tests/test_mixed_modulestore.py

```diff
@@ -741,6 +750,93 @@ def test_publish_automatically_after_delete_unit(self, default_ms):
         assert not self._has_changes(sequential.location)
 
+    @ddt.data(ModuleStoreEnum.Type.mongo, ModuleStoreEnum.Type.split)
+    def test_course_create_event(self, default_ms):
+        """
+        Check that COURSE_CREATED event is sent when a course is created.
+        """
+        self.initdb(default_ms)
+
+        event_receiver = Mock()
+        COURSE_CREATED.connect(event_receiver)
+
+        test_course = self.store.create_course('test_org', 'test_course', 'test_run', self.user_id)
+
+        event_receiver.assert_called()
+
+        self.assertDictContainsSubset(
+            {
+                "signal": COURSE_CREATED,
+                "sender": None,
+                "course": CourseData(
+                    course_key=test_course.id,
+                ),
+            },
+            event_receiver.call_args.kwargs
+        )
+
+    @ddt.data(ModuleStoreEnum.Type.mongo, ModuleStoreEnum.Type.split)
+    def test_xblock_create_event(self, default_ms):
+        """
+        Check that XBLOCK_CREATED event is sent when xblock is created.
+        """
+        self.initdb(default_ms)
+
+        event_receiver = Mock()
+        XBLOCK_CREATED.connect(event_receiver)
+
+        test_course = self.store.create_course('test_org', 'test_course', 'test_run', self.user_id)
+
+        # create sequential test against
+        sequential = self.store.create_child(self.user_id, test_course.location, 'sequential', 'test_sequential')
+
+        event_receiver.assert_called()
+        self.assertDictContainsSubset(
+            {
+                "signal": XBLOCK_CREATED,
+                "sender": None,
+                "xblock_info": XBlockData(
+                    usage_key=sequential.location,
+                    block_type=sequential.location.block_type,
+                ),
+            },
+            event_receiver.call_args.kwargs
+        )
+
+    @ddt.data(ModuleStoreEnum.Type.mongo, ModuleStoreEnum.Type.split)
+    def test_xblock_update_event(self, default_ms):
+        """
+        Check that XBLOCK_UPDATED event is sent when xblock is updated.
+        """
+        self.initdb(default_ms)
+
+        event_receiver = Mock()
+        XBLOCK_UPDATED.connect(event_receiver)
+
+        xblock = self.store.create_item(
+            self.user_id,
+            self.course.id,
+            'vertical',
+            block_id='parent',
+        )
+
+        # Change the xblock
+        xblock.display_name = 'Updated Display Name'
+        self.store.update_item(xblock, user_id=self.user_id)
+
+        event_receiver.assert_called()
+        self.assertDictContainsSubset(
+            {
+                "signal": XBLOCK_UPDATED,
+                "sender": None,
+                "xblock_info": XBlockData(
+                    usage_key=xblock.location,
+                    block_type=xblock.location.block_type,
+                ),
+            },
+            event_receiver.call_args.kwargs
+        )
```

Comment:
@pomegranited Can you give me a hand here? This is generating an error in CI.

The `xblock` variable:

```
XBlockData(usage_key=BlockUsageLocator(CourseLocator('MITx', '999', '2013_Spring', None, None), 'vertical', 'parent'), block_type='vertical')
```

The `call_args.kwgars` from the event:

```
XBlockData(usage_key=BlockUsageLocator(CourseLocator('MITx', '999', '2013_Spring', 'draft-branch', ObjectId('64a5fa4bafc60deeba06e30c')), 'vertical', 'parent'), block_type='vertical')
```
There is a better way to compare these objects?

Reply:
If you can modify the value of `usage_key`, you can use `usage_key = usage_key.for_branch(None)` to strip away the extra branch and version information before you compare it.

It's the branch (`draft-branch`) and version (`ObjectId('64a5fa4bafc60deeba06e30c')`) details embedded in the usage_key which are causing the issue here, but they're not important for this context.

Reply:
Ty @bradenmacdonald!
We should do it for the event data or only to compare in testing?

Reply:
You can do it either way. I don't think the draft/version data is very useful for the events, so you could strip it out from the events. Maybe put the version into `kwargs` in case it's useful to someone. The "branch" will always be `draft-branch` for any edits made in studio, so it's not important information to include.

Reply:
@bradenmacdonald If, when this event is triggered, we need to fetch the updated XBlock content, then don't we need the `version` in order to do that? I'm confused about why we'd want to remove it from the event `usage_key`.

Reply:
@pomegranited At that point it's already saved though, right? So if you retrieve the "current" version, it will include the changes. Or do you want the event handlers to be able to see a "snapshot" of the block as it existed at that time, excluding any future changes which may have been made immediately after, but before the event handler was called? How do other similar events work?

Typically I don't see many people requesting to load a specific XBlock version which is why I didn't think it would be useful in the usage_key. Plus, 99% of the time you see a key, it won't have the version or branch information, so if we leave it in the key, it's a potential footgun for developers who don't realize it's there. Someone might try using that usage_key as an ID, or compiling stats per block, and not realizing that each usage key is always unique (because even for the "same" block, the key will include a different version number) unless you strip out the version to get the "normal" form.

But it is common to want to track when things have changed and to store the version ID somewhere, so I do think it's helpful to have that information available in kwargs.

Reply:
Hi @bradenmacdonald!

As stated [here](https://docs.openedx.org/projects/openedx-events/en/latest/decisions/0003-events-payload.html) the event payload need to "always use the for of attr objects":

> 3. The data sent to an event will always use the form of attr objects following the [OEP-49 data pattern](https://open-edx-proposals.readthedocs.io/en/latest/oep-0049-django-app-patterns.html#id9) . This data objects will be defined in the events library and will be scoped to the Architecture Subdomain they are in. Data objects should use only serializable python primitives and standard types with the exception of OpaqueKeys.

Will we break this pattern if we use `kwarg`?

Also, we don't have a use for the need for both keys right now. Aren't we complicating things in advance?
My vote is for agile simplicity - the art of maximizing the amount of work not done:
We use the better suited key for the actual context (Content authoring auto-tagging). From what you said, I think that is the striped one.

Do you agree, @pomegranited ?

Reply:
@bradenmacdonald 

> At that point it's already saved though, right? So if you retrieve the "current" version, it will include the changes.
> ...
> Plus, 99% of the time you see a key, it won't have the version or branch information, so if we leave it in the key, it's a potential footgun for developers who don't realize it's there.

Ok yep, I understand now.

Content tagging will also be fetching the "current" version, so we don't need the version in the event either.

@rpenido  can you please:

* strip the version out of the UsageKey before sending it with the event
* add an optional `version` string to `XBlockData` , and store it there instead.

> Will we break this pattern if we use kwarg?

Yep, so I think storing the `version` in `XBlockData` gets around that issue.

Reply:
Done!